src/main/java/net/imagej/ops/segment/hough/MidPointAlgorithm.java

```diff
@@ -0,0 +1,363 @@
+	 * @param center
+	 *            the circle center. Must be at least of dimension 2. Dimensions
+	 *            0 and 1 are used to specify the circle center.
+	 * @param radius
+	 *            the circle radius. The circle is written in a plane in
+	 *            dimensions 0 and 1.
+	 * @param value
+	 *            the value to write along the circle.
+	 * @param <T>
+	 *            the type of the target image.
+	 */
+	public static < T extends RealType< T > > void set( final RandomAccessible< T > rai, final Localizable center, final long radius, final T value )
+	{
+		final Cursor< T > cursor = cursor( rai, center, radius );
+		while ( cursor.hasNext() )
+		{
+			cursor.fwd();
+			cursor.get().set( value );
+		}
+	}
+	
+	/**
+	 * Writes a circle in the target {@link RandomAccessible}. The circle is
+	 * written by <b>adding</b> the specified value to the pixel values already
+	 * in the image.
+	 * 
+	 * @param rai
+	 *            the random accessible. It is the caller responsibility to
+	 *            ensure it can be accessed everywhere the circle will be
+	 *            iterated.
+	 * @param center
+	 *            the circle center. Must be at least of dimension 2. Dimensions
+	 *            0 and 1 are used to specify the circle center.
+	 * @param radius
+	 *            the circle radius. The circle is written in a plane in
+	 *            dimensions 0 and 1.
+	 * @param value
+	 *            the value to add along the circle.
+	 * @param <T>
+	 *            the type of the target image.
+	 */
+	public static < T extends RealType< T > > void add( final RandomAccessible< T > rai, final Localizable center, final long radius, final T value )
+	{
+		final Cursor< T > cursor = cursor( rai, center, radius );
+		while ( cursor.hasNext() )
+		{
+			cursor.fwd();
+			cursor.get().add( value );
+		}
+	}
+	
+	private MidPointAlgorithm()
+	{}
+
+	private static final class MidPointCursor< T > implements Cursor< T >
+	{
+
+		private final RandomAccessible< T > rai;
+
+		private final RandomAccess< T > ra;
+
+		private final Localizable center;
+
+		private final long radius;
+
+		private final int dimX;
+
+		private final int dimY;
+
+		private long x;
+
+		private long y;
+
+		private long dx;
+
+		private long dy;
+
+		private long f;
+
+		private Octant octant;
+
+		private boolean hasNext;
+
+
+		private static enum Octant
+		{
+			INIT, EAST, NORTH, WEST, SOUTH, O1, O2, O3, O4, O5, O6, O7, O8;
+		}
+
+		public MidPointCursor( final RandomAccessible< T > rai, final Localizable center, final long radius, final int dimX, final int dimY )
+		{
+			this.rai = rai;
+			this.center = center;
+			this.radius = radius;
+			this.dimX = dimX;
+			this.dimY = dimY;
+			this.ra = rai.randomAccess();
+			reset();
+		}
+
+		@Override
+		public void reset()
+		{
+			x = 0;
+			y = radius;
+			f = 1 - radius;
+			dx = 1;
+			dy = -2 * radius;
+			octant = Octant.INIT;
+			ra.setPosition( center );
+			hasNext = true;
+		}
+
+		@Override
+		public void fwd()
```

Comment:
I am no expert on time, but it seems to me like it would be much easier and maybe even faster to, at every radius, do this just once to create a lookup table, and then just access the lookup table every time this is called. Would that not make sense here?

Reply:
Normally this is how circles are painted on the screen (https://en.wikipedia.org/wiki/Midpoint_circle_algorithm). It is supposed to be very efficient and can deal with whatever radius we pick, without too much memory allocation.